source/auth/auth.rst

```diff
@@ -1175,70 +1175,146 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
 
+
 MONGODB-OIDC
 ~~~~~~~~~~~~
 
 :since: 7.0 Enterprise
 
-MONGODB-OIDC authenticates using an `OIDC <https://openid.net/specs/openid-connect-core-1_0.html>`_ access tokens.  Drivers MUST support
-both Callback-driven OIDC and Automatic OIDC Authentication for AWS.
+MONGODB-OIDC authenticates using an `OpenID Connect (OIDC)
+<https://openid.net/specs/openid-connect-core-1_0.html>`_ access token.
+
+TODO: Add background on what part of the auth flow drivers play specifically.
+
+`MongoCredential`_ Properties
+`````````````````````````````
+
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
 
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PROVIDER_NAME
+        Drivers MUST allow the user to specify the name of the OIDC service to
+        use to obtain credentials. MUST be one of ["aws"]. If PROVIDER_NAME is
+        given and a Custom provider callback is configured, the driver MUST
+        raise an error.
+
+Supported Service Providers
+```````````````````````````
+
+AWS
+___
+
+TODO: Add.
+
+Custom
+______
+
+Drivers MUST allow the user to integrate with other OIDC providers by
+implementing a custom callback that returns an OIDC token. Drivers MUST provide
+a way for the callback to be either automatically canceled, or to cancel itself.
+This can be as a timeout argument to the callback, a cancellation context passed
+to the callback, or some other language-appropriate mechanism.  The timeout
+deadline MUST be the same as the connection establishment deadline.
+
+Callbacks can be synchronous and/or asynchronous, depending on the driver
+and/or language.  Asynchronous callbacks should be preferred when other
+operations in the driver use asynchronous functions.
+
+The driver MUST pass the following information to the callback:
+
+- ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.
+- ``mechanismProperties``: All key-value pairs from the
+  AUTH_MECHANISM_PROPERTIES parameter.
+- ``invalidatedToken``: Optional. The previous token that was invalidated after
+  receiving a ``ReauthenticationRequired`` error.
+
+The callback MUST return an OIDC access token in JWT format.
+
+The signature of the callback is up to the driver's discretion, but the driver
+MUST ensure that additional optional parameters can be added to the callback
+signature in the future. An example might look like:
+
+.. code:: typescript
+
+  interface CallbackParameters {
+      // All key-value pairs from the AUTH_MECHANISM_PROPERTIES parameter.
+      mechanismProperties: Object;
+
+      // The refresh token, if applicable, to be used by the callback to request a new token from the issuer.
+      invalidatedToken: Optional<string>;
```

Comment:
I think this comment is incorrect. The invalidated token is the "previous" accessToken (that is, the accessToken associated with the connection that has received a 391-reauth).

The user, who is expected to handle caching and locking on their end, would use this to see if the cache is invalidated (that is, if the cache's token is the same as this invalid token, the cached refreshToken should then be used to communicate with the issuer).

We might consider renaming this to `invalidatedAccessToken`, since this might be confusing for users as well.

Reply:
I believe there's no longer a reason to pass the invalidated token into the callback because it was only needed to allow the callback author to manage their cache. Since we're handling caching in the driver, it shouldn't be necessary, so I'm planning to remove it.

Reply:
Removed.